Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp

```diff
@@ -194,7 +201,58 @@ struct HuffmanTable {
                 codes.append(code++);
             code <<= 1;
         }
+
+        generate_lookup_table();
     }
+
+    struct SymbolAndSize {
+        u8 symbol {};
+        u8 size {};
+    };
+
+    ErrorOr<SymbolAndSize> symbol_from_code(u16 code) const
+    {
+        static constexpr u8 shift_for_cache = maximum_bits_per_code - bits_per_cached_code;
+
+        if (lookup_table[code >> shift_for_cache] != invalid_entry) {
+            u8 const code_length = lookup_table[code >> shift_for_cache] >> bits_per_cached_code;
+            return SymbolAndSize { (u8)lookup_table[code >> shift_for_cache], code_length };
+        }
+
+        u64 code_cursor = first_non_cached_code_index;
+
+        for (u8 i = HuffmanTable::bits_per_cached_code; i < 16; i++) {
+            auto const result = code >> (maximum_bits_per_code - 1 - i);
+            for (u32 j = 0; j < code_counts[i]; j++) {
+                if (result == codes[code_cursor])
+                    return SymbolAndSize { symbols[code_cursor], (u8)(i + 1) };
+
+                code_cursor++;
+            }
+        }
+
+        return Error::from_string_literal("This kind of JPEG is not yet supported by the decoder");
+    }
+
+private:
+    void generate_lookup_table()
+    {
+        lookup_table.fill(invalid_entry);
+
+        u32 code_offset = 0;
+        for (u8 code_length = 1; code_length <= bits_per_cached_code; code_length++) {
+            for (u32 i = 0; i < code_counts[code_length - 1]; i++, code_offset++) {
+                u32 code_key = codes[code_offset] << (bits_per_cached_code - code_length);
+                for (u8 duplicate_count = 1 << (bits_per_cached_code - code_length); duplicate_count > 0; duplicate_count--) {
+                    lookup_table[code_key] = (code_length << bits_per_cached_code) | symbols[code_offset];
+                    code_key++;
+                }
+            }
+        }
+    }
+
+    static constexpr u16 invalid_entry = 0xFF;
```

Comment:
In our usual ordering, these statics come before the methods inside of this visibility scope (i.e. just after `private:`)